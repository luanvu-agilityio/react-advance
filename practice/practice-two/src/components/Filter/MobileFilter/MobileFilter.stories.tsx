import type { Meta, StoryObj } from '@storybook/react'
import { useState, type ChangeEvent } from 'react'
import { MobileFilter } from './MobileFilter'
import type {
  CategoryProps,
  BrandProps,
  RatingProps,
  PriceRangeProps,
} from 'types/Filter'

const meta: Meta<typeof MobileFilter> = {
  title: 'Components/Filter/MobileFilter',
  component: MobileFilter,
  parameters: {
    layout: 'centered',
    backgrounds: {
      default: 'light',
    },
    docs: {
      description: {
        component:
          'Mobile filter component with dialog interface for filtering products on mobile devices.',
      },
    },
  },
  tags: ['autodocs'],
  argTypes: {
    activeCategory: {
      control: 'text',
      description: 'Currently active category name',
    },
    sliderValues: {
      control: 'object',
      description: 'Current slider values for price range',
    },
    onCategoryClick: {
      action: 'category-clicked',
      description: 'Callback when a category is clicked',
    },
    onBrandChange: {
      action: 'brand-changed',
      description: 'Callback when a brand selection changes',
    },
    onRatingChange: {
      action: 'rating-changed',
      description: 'Callback when a rating selection changes',
    },
    onSliderChange: {
      action: 'slider-changed',
      description: 'Callback when price slider changes',
    },
    onApplyFilters: {
      action: 'filters-applied',
      description: 'Callback when apply filters button is clicked',
    },
    onResetFilters: {
      action: 'filters-reset',
      description: 'Callback when reset filters button is clicked',
    },
  },
}

export default meta
type Story = StoryObj<typeof MobileFilter>

// Sample data for filters
const sampleCategories: CategoryProps[] = [
  { name: 'Bakery', count: 12 },
  { name: 'Fruits', count: 24 },
  { name: 'Vegetables', count: 32 },
  { name: 'Dairy', count: 18 },
  { name: 'Meat', count: 16 },
]

const sampleBrands: BrandProps[] = [
  { name: 'United Farms', selected: false, count: 15 },
  { name: 'Organic Farms', selected: true, count: 12 },
  { name: 'Fresh Fields', selected: false, count: 8 },
  { name: 'Green Gardens', selected: false, count: 10 },
  { name: "Nature's Best", selected: true, count: 20 },
]

const sampleRatings: RatingProps[] = [
  { rating: 5, selected: false },
  { rating: 4, selected: true },
  { rating: 3, selected: false },
  { rating: 2, selected: false },
  { rating: 1, selected: false },
]

const samplePriceRange: PriceRangeProps = { min: 100, max: 500 }

/**
 * Default MobileFilter with pre-configured filter options
 */
export const Default: Story = {
  args: {
    categories: sampleCategories,
    brands: sampleBrands,
    ratings: sampleRatings,
    priceRange: samplePriceRange,
    activeCategory: 'Fruits',
    sliderValues: [100, 500],
    onCategoryClick: (categoryName) => {
      console.log(`Category clicked: ${categoryName}`)
    },
    onBrandChange: (index) => {
      console.log(`Brand at index ${index} clicked`)
    },
    onRatingChange: (index) => {
      console.log(`Rating at index ${index} clicked`)
    },
    onSliderChange: (values) => {
      console.log(`Slider values changed: ${values}`)
    },
    onMinInputChange: (e) => {
      console.log(`Min input changed: ${e.target.value}`)
    },
    onMaxInputChange: (e) => {
      console.log(`Max input changed: ${e.target.value}`)
    },
    onApplyFilters: () => {
      console.log('Apply filters clicked')
    },
    onResetFilters: () => {
      console.log('Reset filters clicked')
    },
    // Add the missing optional props
    onBrandSelect: (brandName, isSelected) => {
      console.log(`Brand ${brandName} selected: ${isSelected}`)
    },
    onRatingSelect: (rating, isSelected) => {
      console.log(`Rating ${rating} selected: ${isSelected}`)
    },
    onPriceRangeChange: (min, max) => {
      console.log(`Price range changed: ${min} - ${max}`)
    },
  },
}

/**
 * Interactive MobileFilter that demonstrates state changes
 */
export const Interactive: Story = {
  render: function InteractiveMobileFilter() {
    // State for all filter options
    const [activeCategory, setActiveCategory] = useState('Fruits')
    const [brands, setBrands] = useState<BrandProps[]>(sampleBrands)
    const [ratings, setRatings] = useState<RatingProps[]>(sampleRatings)
    const [priceRange, setPriceRange] =
      useState<PriceRangeProps>(samplePriceRange)
    const [sliderValues, setSliderValues] = useState([100, 500])

    // Handler functions
    const handleCategoryClick = (categoryName: string) => {
      const newActiveCategory =
        categoryName === activeCategory ? '' : categoryName
      setActiveCategory(newActiveCategory)
      console.log('Category clicked:', newActiveCategory)
    }

    const handleBrandChange = (index: number) => {
      const newBrands = [...brands]
      newBrands[index].selected = !newBrands[index].selected
      setBrands(newBrands)
      console.log('Brand changed:', newBrands[index])
    }

    const handleRatingChange = (index: number) => {
      const newRatings = [...ratings]
      newRatings[index].selected = !newRatings[index].selected
      setRatings(newRatings)
      console.log('Rating changed:', newRatings[index])
    }

    const handleSliderChange = (values: number[]) => {
      setSliderValues(values)
      setPriceRange({
        min: values[0],
        max: values[1],
      })
      console.log('Slider changed:', values)
    }

    const handleMinInputChange = (e: ChangeEvent<HTMLInputElement>) => {
      const value = Math.min(Number(e.target.value), priceRange.max)
      setPriceRange({
        ...priceRange,
        min: value,
      })
      setSliderValues([value, sliderValues[1]])
    }

    const handleMaxInputChange = (e: ChangeEvent<HTMLInputElement>) => {
      const value = Math.max(Number(e.target.value), priceRange.min)
      setPriceRange({
        ...priceRange,
        max: value,
      })
      setSliderValues([sliderValues[0], value])
    }

    const handleApplyFilters = () => {
      const appliedFilters = {
        activeCategory,
        brands: brands.filter((b) => b.selected).map((b) => b.name),
        ratings: ratings.filter((r) => r.selected).map((r) => r.rating),
        priceRange,
      }
      console.log('Filters applied:', appliedFilters)
      alert(
        `Filters applied! Selected ${appliedFilters.brands.length} brands, ${appliedFilters.ratings.length} ratings. Check console for details.`
      )
    }

    const handleResetFilters = () => {
      setActiveCategory('')
      setBrands(sampleBrands.map((b) => ({ ...b, selected: false })))
      setRatings(sampleRatings.map((r) => ({ ...r, selected: false })))
      setPriceRange({ min: 0, max: 1000 })
      setSliderValues([0, 1000])
      console.log('Filters reset')
      alert('All filters have been reset!')
    }

    // Additional handlers for alternative prop signatures
    const handleBrandSelect = (brandName: string, isSelected: boolean) => {
      const brandIndex = brands.findIndex((brand) => brand.name === brandName)
      if (brandIndex !== -1) {
        const newBrands = [...brands]
        newBrands[brandIndex].selected = isSelected
        setBrands(newBrands)
        console.log(
          `Brand ${brandName} ${isSelected ? 'selected' : 'deselected'}`
        )
      }
    }

    const handleRatingSelect = (rating: number, isSelected: boolean) => {
      const ratingIndex = ratings.findIndex((r) => r.rating === rating)
      if (ratingIndex !== -1) {
        const newRatings = [...ratings]
        newRatings[ratingIndex].selected = isSelected
        setRatings(newRatings)
        console.log(
          `Rating ${rating} ${isSelected ? 'selected' : 'deselected'}`
        )
      }
    }

    const handlePriceRangeChange = (minPrice: number, maxPrice: number) => {
      setPriceRange({ min: minPrice, max: maxPrice })
      setSliderValues([minPrice, maxPrice])
      console.log(`Price range changed: ${minPrice} - ${maxPrice}`)
    }

    return (
      <div style={{ width: '100%', maxWidth: '375px', margin: '0 auto' }}>
        <MobileFilter
          categories={sampleCategories}
          brands={brands}
          ratings={ratings}
          priceRange={priceRange}
          activeCategory={activeCategory}
          sliderValues={sliderValues}
          onCategoryClick={handleCategoryClick}
          onBrandChange={handleBrandChange}
          onRatingChange={handleRatingChange}
          onSliderChange={handleSliderChange}
          onMinInputChange={handleMinInputChange}
          onMaxInputChange={handleMaxInputChange}
          onApplyFilters={handleApplyFilters}
          onResetFilters={handleResetFilters}
          onBrandSelect={handleBrandSelect}
          onRatingSelect={handleRatingSelect}
          onPriceRangeChange={handlePriceRangeChange}
        />
        <div
          style={{
            marginTop: 20,
            padding: 16,
            backgroundColor: '#f5f5f5',
            borderRadius: 8,
            fontSize: '14px',
            lineHeight: '1.4',
          }}
        >
          <p style={{ margin: '0 0 8px 0', fontWeight: 'bold' }}>
            ðŸ“± Interactive Demo
          </p>
          <p style={{ margin: '0 0 4px 0' }}>
            â€¢ Click the "Filters" button above to open the mobile filter dialog
          </p>
          <p style={{ margin: '0 0 4px 0' }}>
            â€¢ All interactions will update the state and log to console
          </p>
          <p style={{ margin: 0 }}>
            â€¢ Try selecting different categories, brands, ratings, and price
            ranges
          </p>
        </div>
      </div>
    )
  },
  parameters: {
    docs: {
      description: {
        story:
          'Interactive version with full state management. Open the filter dialog and try different combinations.',
      },
    },
  },
}

/**
 * MobileFilter with minimal options
 */
export const MinimalOptions: Story = {
  args: {
    categories: [{ name: 'All Products', count: 120 }],
    brands: [{ name: 'The Only Brand', selected: false, count: 1 }],
    ratings: [{ rating: 5, selected: false }],
    priceRange: { min: 0, max: 100 },
    activeCategory: '',
    sliderValues: [0, 100],
    onCategoryClick: () => {},
    onBrandChange: () => {},
    onRatingChange: () => {},
    onSliderChange: () => {},
    onMinInputChange: () => {},
    onMaxInputChange: () => {},
    onApplyFilters: () => {},
    onResetFilters: () => {},
    onBrandSelect: () => {},
    onRatingSelect: () => {},
    onPriceRangeChange: () => {},
  },
  parameters: {
    docs: {
      description: {
        story:
          'Minimal configuration with just one option in each filter category.',
      },
    },
  },
}

/**
 * MobileFilter with lots of filter options
 */
export const ManyOptions: Story = {
  args: {
    categories: Array(15)
      .fill(null)
      .map((_, i) => ({
        name: `Category ${i + 1}`,
        count: Math.floor(Math.random() * 50) + 5,
      })),
    brands: Array(20)
      .fill(null)
      .map((_, i) => ({
        name: `Brand ${i + 1}`,
        selected: i % 5 === 0, // Every 5th brand is selected
        count: Math.floor(Math.random() * 50) + 5,
      })),
    ratings: [
      { rating: 5, selected: true },
      { rating: 4, selected: true },
      { rating: 3, selected: false },
      { rating: 2, selected: false },
      { rating: 1, selected: false },
    ],
    priceRange: { min: 25, max: 750 },
    activeCategory: 'Category 5',
    sliderValues: [25, 750],
    onCategoryClick: () => {},
    onBrandChange: () => {},
    onRatingChange: () => {},
    onSliderChange: () => {},
    onMinInputChange: () => {},
    onMaxInputChange: () => {},
    onApplyFilters: () => {},
    onResetFilters: () => {},
    onBrandSelect: () => {},
    onRatingSelect: () => {},
    onPriceRangeChange: () => {},
  },
  parameters: {
    docs: {
      description: {
        story:
          'Example with many filter options to test scrolling and performance with large datasets.',
      },
    },
  },
}

/**
 * Empty state when no filter options are available
 */
export const EmptyState: Story = {
  args: {
    categories: [],
    brands: [],
    ratings: [],
    priceRange: { min: 0, max: 0 },
    activeCategory: '',
    sliderValues: [0, 0],
    onCategoryClick: () => {},
    onBrandChange: () => {},
    onRatingChange: () => {},
    onSliderChange: () => {},
    onMinInputChange: () => {},
    onMaxInputChange: () => {},
    onApplyFilters: () => {},
    onResetFilters: () => {},
    onBrandSelect: () => {},
    onRatingSelect: () => {},
    onPriceRangeChange: () => {},
  },
  parameters: {
    docs: {
      description: {
        story: 'Empty state when no filter options are available.',
      },
    },
  },
}
