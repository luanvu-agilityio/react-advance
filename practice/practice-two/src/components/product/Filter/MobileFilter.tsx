import { Filter, X } from 'lucide-react'
import * as Dialog from '@radix-ui/react-dialog'
import {
  MobileFilter as MobileFilterWrapper,
  MobileFilterButton,
  DialogOverlay,
  DialogContent,
  DialogHeader,
  DialogTitle,
  CloseButton,
  FilterContainer,
} from './FilterStyles'
import { CategoryFilter } from './CategoryFilter'
import { BrandFilter } from './BrandFilter'
import { RatingFilter } from './RatingFilter'
import { PriceFilter } from './PriceFilter'
import type { FilterComponentsProps } from 'types/Filter'
import type { ChangeEvent } from 'react'

interface MobileFilterProps
  extends Omit<FilterComponentsProps, 'initialActiveCategory'> {
  activeCategory: string
  sliderValues: number[]
  onCategoryClick: (categoryName: string) => void
  onBrandChange: (index: number) => void
  onRatingChange: (index: number) => void
  onSliderChange: (values: number[]) => void
  onMinInputChange: (e: ChangeEvent<HTMLInputElement>) => void
  onMaxInputChange: (e: ChangeEvent<HTMLInputElement>) => void
  onApplyFilters: () => void
  onResetFilters: () => void
}

export const MobileFilter = ({
  categories,
  brands,
  ratings,
  priceRange,
  activeCategory,
  sliderValues,
  onCategoryClick,
  onBrandChange,
  onRatingChange,
  onSliderChange,
  onMinInputChange,
  onMaxInputChange,
  onApplyFilters,
  onResetFilters,
}: MobileFilterProps) => {
  return (
    <MobileFilterWrapper>
      <Dialog.Root>
        <Dialog.Trigger asChild>
          <MobileFilterButton>
            <Filter size={16} />
            Filters
          </MobileFilterButton>
        </Dialog.Trigger>

        <Dialog.Portal>
          <DialogOverlay />
          <DialogContent>
            <DialogHeader>
              <DialogTitle>Filters</DialogTitle>
              <CloseButton>
                <X size={20} />
              </CloseButton>
            </DialogHeader>

            <FilterContainer>
              <CategoryFilter
                categories={categories}
                activeCategory={activeCategory}
                onCategoryClick={onCategoryClick}
              />

              <BrandFilter brands={brands} onBrandChange={onBrandChange} />

              <RatingFilter ratings={ratings} onRatingChange={onRatingChange} />

              <PriceFilter
                priceRange={priceRange}
                sliderValues={sliderValues}
                onSliderChange={onSliderChange}
                onMinInputChange={onMinInputChange}
                onMaxInputChange={onMaxInputChange}
                onApplyFilters={onApplyFilters}
                onResetFilters={onResetFilters}
              />
            </FilterContainer>
          </DialogContent>
        </Dialog.Portal>
      </Dialog.Root>
    </MobileFilterWrapper>
  )
}
