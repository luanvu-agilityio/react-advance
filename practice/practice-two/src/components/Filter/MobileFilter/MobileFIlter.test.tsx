import { render, screen, fireEvent } from '@testing-library/react'
import { MobileFilter } from './MobileFilter'
import type {
  CategoryProps,
  BrandProps,
  RatingProps,
  PriceRangeProps,
} from 'types/Filter'
import type { ReactNode } from 'react'

// Mock the Dialog component from Radix
jest.mock('@radix-ui/react-dialog', () => {
  const originalModule = jest.requireActual('@radix-ui/react-dialog')
  return {
    __esModule: true,
    ...originalModule,
    Root: ({ children }: { children: ReactNode }) => (
      <div data-testid="dialog-root">{children}</div>
    ),
    Trigger: ({ children }: { children: ReactNode }) => (
      <button data-testid="dialog-trigger">{children}</button>
    ),
    Portal: ({ children }: { children: ReactNode }) => (
      <div data-testid="dialog-portal">{children}</div>
    ),
    Content: ({ children }: { children: ReactNode }) => (
      <div data-testid="dialog-content">{children}</div>
    ),
    Close: ({ children }: { children: ReactNode }) => (
      <button data-testid="dialog-close">{children}</button>
    ),
  }
})

// Mock child filter components
jest.mock('./CategoryFilter/CategoryFilter', () => ({
  CategoryFilter: ({
    categories,
    activeCategory,
    onCategoryClick,
  }: {
    categories: CategoryProps[]
    activeCategory: string
    onCategoryClick: (category: string) => void
  }) => (
    <div data-testid="category-filter">
      <span>Categories: {categories.length}</span>
      <span>Active: {activeCategory}</span>
      <button onClick={() => onCategoryClick('Vegetables')}>
        Click Category
      </button>
    </div>
  ),
}))

jest.mock('./BrandFIlter/BrandFilter', () => ({
  BrandFilter: ({
    brands,
    onBrandChange,
  }: {
    brands: BrandProps[]
    onBrandChange: (index: number) => void
  }) => (
    <div data-testid="brand-filter">
      <span>Brands: {brands.length}</span>
      <button onClick={() => onBrandChange(0)}>Click Brand</button>
    </div>
  ),
}))

jest.mock('./RatingFilter/RatingFilter', () => ({
  RatingFilter: ({
    ratings,
    onRatingChange,
  }: {
    ratings: RatingProps[]
    onRatingChange: (index: number) => void
  }) => (
    <div data-testid="rating-filter">
      <span>Ratings: {ratings.length}</span>
      <button onClick={() => onRatingChange(0)}>Click Rating</button>
    </div>
  ),
}))

jest.mock('./PriceFilter/PriceFilter', () => ({
  PriceFilter: ({
    priceRange,
    sliderValues,
    onSliderChange,
    onMinInputChange,
    onMaxInputChange,
    onApplyFilters,
    onResetFilters,
  }: {
    priceRange: PriceRangeProps
    sliderValues: number[]
    onSliderChange: (values: number[]) => void
    onMinInputChange: (e: { target: { value: string } }) => void
    onMaxInputChange: (e: { target: { value: string } }) => void
    onApplyFilters: () => void
    onResetFilters: () => void
  }) => (
    <div data-testid="price-filter">
      <span>
        Price: ${priceRange.min} - ${priceRange.max}
      </span>
      <span>
        Slider: ${sliderValues[0]} - ${sliderValues[1]}
      </span>
      <button onClick={() => onSliderChange([200, 300])}>Change Slider</button>
      <button onClick={() => onMinInputChange({ target: { value: '150' } })}>
        Change Min
      </button>
      <button onClick={() => onMaxInputChange({ target: { value: '450' } })}>
        Change Max
      </button>
      <button onClick={onApplyFilters}>Apply</button>
      <button onClick={onResetFilters}>Reset</button>
    </div>
  ),
}))

// Mock Lucide icons
jest.mock('lucide-react', () => ({
  Filter: () => <div data-testid="filter-icon">Filter Icon</div>,
  X: () => <div data-testid="close-icon">X Icon</div>,
}))

describe('MobileFilter', () => {
  // Sample filter data for tests
  const sampleCategories: CategoryProps[] = [
    { name: 'Bakery', count: 12 },
    { name: 'Fruits', count: 24 },
    { name: 'Vegetables', count: 32 },
  ]

  const sampleBrands: BrandProps[] = [
    { name: 'United Farms', selected: false },
    { name: 'Organic Farms', selected: true },
  ]

  const sampleRatings: RatingProps[] = [
    { rating: 5, selected: false },
    { rating: 4, selected: true },
  ]

  const samplePriceRange: PriceRangeProps = { min: 100, max: 500 }

  const mockProps = {
    categories: sampleCategories,
    brands: sampleBrands,
    ratings: sampleRatings,
    priceRange: samplePriceRange,
    activeCategory: 'Fruits',
    sliderValues: [100, 500],
    onCategoryClick: jest.fn(),
    onBrandChange: jest.fn(),
    onRatingChange: jest.fn(),
    onSliderChange: jest.fn(),
    onMinInputChange: jest.fn(),
    onMaxInputChange: jest.fn(),
    onApplyFilters: jest.fn(),
    onResetFilters: jest.fn(),
    onBrandSelect: jest.fn(),
    onRatingSelect: jest.fn(),
    onPriceRangeChange: jest.fn(),
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders the mobile filter button correctly', () => {
    render(<MobileFilter {...mockProps} />)

    const filterButton = screen.getByTestId('dialog-trigger')
    expect(filterButton).toBeInTheDocument()
    expect(screen.getByTestId('filter-icon')).toBeInTheDocument()
    expect(filterButton).toHaveTextContent('Filters')
  })

  it('renders filter components inside dialog content', () => {
    render(<MobileFilter {...mockProps} />)

    // Dialog content should be rendered (even if not visible in actual DOM)
    expect(screen.getByTestId('dialog-content')).toBeInTheDocument()

    // Check that all filter components are rendered
    expect(screen.getByTestId('category-filter')).toBeInTheDocument()
    expect(screen.getByTestId('brand-filter')).toBeInTheDocument()
    expect(screen.getByTestId('rating-filter')).toBeInTheDocument()
    expect(screen.getByTestId('price-filter')).toBeInTheDocument()
  })

  it('passes correct props to child filter components', () => {
    render(<MobileFilter {...mockProps} />)

    // Check that CategoryFilter received correct props
    expect(screen.getByText('Categories: 3')).toBeInTheDocument()
    expect(screen.getByText('Active: Fruits')).toBeInTheDocument()

    // Check that BrandFilter received correct props
    expect(screen.getByText('Brands: 2')).toBeInTheDocument()

    // Check that RatingFilter received correct props
    expect(screen.getByText('Ratings: 2')).toBeInTheDocument()

    // Check that PriceFilter received correct props
    expect(screen.getByText('Price: $100 - $500')).toBeInTheDocument()
    expect(screen.getByText('Slider: $100 - $500')).toBeInTheDocument()
  })

  it('calls category click handler when category is clicked', () => {
    render(<MobileFilter {...mockProps} />)

    // Click the category button in the mocked CategoryFilter
    fireEvent.click(screen.getByText('Click Category'))

    // Check that the onCategoryClick handler was called with the correct argument
    expect(mockProps.onCategoryClick).toHaveBeenCalledTimes(1)
    expect(mockProps.onCategoryClick).toHaveBeenCalledWith('Vegetables')
  })

  it('calls brand change handler when brand is clicked', () => {
    render(<MobileFilter {...mockProps} />)

    // Click the brand button in the mocked BrandFilter
    fireEvent.click(screen.getByText('Click Brand'))

    // Check that the onBrandChange handler was called with the correct argument
    expect(mockProps.onBrandChange).toHaveBeenCalledTimes(1)
    expect(mockProps.onBrandChange).toHaveBeenCalledWith(0)
  })

  it('calls rating change handler when rating is clicked', () => {
    render(<MobileFilter {...mockProps} />)

    // Click the rating button in the mocked RatingFilter
    fireEvent.click(screen.getByText('Click Rating'))

    // Check that the onRatingChange handler was called with the correct argument
    expect(mockProps.onRatingChange).toHaveBeenCalledTimes(1)
    expect(mockProps.onRatingChange).toHaveBeenCalledWith(0)
  })

  it('calls price filter handlers when price filter actions are taken', () => {
    render(<MobileFilter {...mockProps} />)

    // Click the slider change button in the mocked PriceFilter
    fireEvent.click(screen.getByText('Change Slider'))

    // Check that the onSliderChange handler was called with the correct argument
    expect(mockProps.onSliderChange).toHaveBeenCalledTimes(1)
    expect(mockProps.onSliderChange).toHaveBeenCalledWith([200, 300])

    // Click the min input change button
    fireEvent.click(screen.getByText('Change Min'))

    // Check that the onMinInputChange handler was called with the correct argument
    expect(mockProps.onMinInputChange).toHaveBeenCalledTimes(1)
    expect(mockProps.onMinInputChange).toHaveBeenCalledWith(
      expect.objectContaining({
        target: { value: '150' },
      })
    )

    // Click the max input change button
    fireEvent.click(screen.getByText('Change Max'))

    // Check that the onMaxInputChange handler was called with the correct argument
    expect(mockProps.onMaxInputChange).toHaveBeenCalledTimes(1)
    expect(mockProps.onMaxInputChange).toHaveBeenCalledWith(
      expect.objectContaining({
        target: { value: '450' },
      })
    )

    // Click the apply button
    fireEvent.click(screen.getByText('Apply'))

    // Check that the onApplyFilters handler was called
    expect(mockProps.onApplyFilters).toHaveBeenCalledTimes(1)

    // Click the reset button
    fireEvent.click(screen.getByText('Reset'))

    // Check that the onResetFilters handler was called
    expect(mockProps.onResetFilters).toHaveBeenCalledTimes(1)
  })

  it('matches snapshot', () => {
    const { asFragment } = render(<MobileFilter {...mockProps} />)
    expect(asFragment()).toMatchSnapshot()
  })
})
