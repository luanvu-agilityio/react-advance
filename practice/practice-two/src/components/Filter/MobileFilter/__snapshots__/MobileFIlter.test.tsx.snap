// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileFilter matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-kiMgGE eQcaTe"
  >
    <div
      data-testid="dialog-root"
    >
      <button
        data-testid="dialog-trigger"
      />
      <button
        class="sc-dTUlgT dbZUsC"
      >
        <div
          data-testid="filter-icon"
        >
          Filter Icon
        </div>
        Filters
      </button>
      <div
        data-testid="dialog-portal"
      >
        <div
          data-testid="dialog-overlay"
        />
        <div
          data-testid="dialog-content"
        >
          <div
            class="sc-iSqusk ctiGHA"
          >
            <h2
              data-testid="dialog-title"
            >
              Filters
            </h2>
            <button
              data-testid="dialog-close"
            >
              <div
                data-testid="close-icon"
              >
                X Icon
              </div>
            </button>
          </div>
          <div
            class="sc-Qotzb DsboM"
          >
            <div
              data-testid="category-filter"
            >
              <span>
                Categories: 3
              </span>
              <span>
                Active: Fruits
              </span>
              <button>
                Click Category
              </button>
            </div>
            <div
              data-testid="brand-filter"
            >
              <span>
                Brands: 2
              </span>
              <button>
                Click Brand
              </button>
            </div>
            <div
              data-testid="rating-filter"
            >
              <span>
                Ratings: 2
              </span>
              <button>
                Click Rating
              </button>
            </div>
            <div
              data-testid="price-filter"
            >
              <span>
                Price: $100 - $500
              </span>
              <span>
                Slider: $100 - $500
              </span>
              <button>
                Change Slider
              </button>
              <button>
                Change Min
              </button>
              <button>
                Change Max
              </button>
              <button>
                Apply
              </button>
              <button>
                Reset
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
